core/vm/ipgraph_with_kind.go

```diff
@@ -0,0 +1,365 @@
+package vm
```

Comment:
While the computations do keep track of processed nodes in the ancestors map, it may not be preventing revisiting nodes. In a cyclic graph, this could lead to infinite loops?

So the question is if the licensing related restrictions are adequately preventing cycles from being created via that `addParentIp` precompile.

Examples of cycles we may want to avoid and confirm are not possible to be made due to protocol restrictions:
Example 1:
A is parent of B
B is parent of A

Example 2:
A is parent of B
B is parent of C
C is parent of A

Reply:
yes, on smart contract we prevent cyclic graph, the ip cannot link to any new parents, when the ip already has parents or children.